In [1]:
import pandas as pd

In [2]:
data_path = '201902-fordgobike-tripdata.csv'
df = pd.read_csv(data_path)

In [3]:
print(df.shape,"\n")
print(df.dtypes,"\n")
display(df.head(5))

(183412, 16) 

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id           float64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year          float64
member_gender               object
bike_share_for_all_trip     object
dtype: object 



duration_sec                start_time                  end_time   
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750  \
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   
2         61854  2019-02-28 12:13:13.2180  2019-03-01 05:24:08.1460   
3         36490  2019-02-28 17:54:26.0100  2019-03-01 04:02:36.8420   
4          1585  2019-02-28 23:54:18.5490  2019-03-01 00:20:44.0740   

   start_station_id                                start_station_name   
0              21.0  Montgomery St BART Station (Market St at 2nd St)  \
1              23.0                     The Embarcadero at Steuart St   
2              86.0                           Market St at Dolores St   
3             375.0                           Grove St at Masonic Ave   
4               7.0                               Frank H Ogawa Plaza   

   start_station_latitude  start_station_longitude  end_station_id   
0               37.789625              -122.400811            13.0  \
1               37.791464              -122.391034            81.0   
2               37.769305              -122.426826             3.0   
3               37.774836              -122.446546            70.0   
4               37.804562              -122.271738           222.0   

                               end_station_name  end_station_latitude   
0                Commercial St at Montgomery St             37.794231  \
1                            Berry St at 4th St             37.775880   
2  Powell St BART Station (Market St at 4th St)             37.786375   
3                        Central Ave at Fell St             37.773311   
4                         10th Ave at E 15th St             37.792714   

   end_station_longitude  bike_id   user_type  member_birth_year   
0            -122.402923     4902    Customer             1984.0  \
1            -122.393170     2535    Customer                NaN   
2            -122.404904     5905    Customer             1972.0   
3            -122.444293     6638  Subscriber             1989.0   
4            -122.248780     4898  Subscriber             1974.0   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1           NaN                      No  
2          Male                      No  
3         Other                      No  
4          Male                     Yes

In [4]:
# descriptive statistics for numeric variables
display(df.describe())

duration_sec  start_station_id  start_station_latitude   
count  183412.000000     183215.000000           183412.000000  \
mean      726.078435        138.590427               37.771223   
std      1794.389780        111.778864                0.099581   
min        61.000000          3.000000               37.317298   
25%       325.000000         47.000000               37.770083   
50%       514.000000        104.000000               37.780760   
75%       796.000000        239.000000               37.797280   
max     85444.000000        398.000000               37.880222   

       start_station_longitude  end_station_id  end_station_latitude   
count            183412.000000   183215.000000         183412.000000  \
mean               -122.352664      136.249123             37.771427   
std                   0.117097      111.515131              0.099490   
min                -122.453704        3.000000             37.317298   
25%                -122.412408       44.000000             37.770407   
50%                -122.398285      100.000000             37.781010   
75%                -122.286533      235.000000             37.797320   
max                -121.874119      398.000000             37.880222   

       end_station_longitude        bike_id  member_birth_year  
count          183412.000000  183412.000000      175147.000000  
mean             -122.352250    4472.906375        1984.806437  
std                 0.116673    1664.383394          10.116689  
min              -122.453704      11.000000        1878.000000  
25%              -122.411726    3777.000000        1980.000000  
50%              -122.398279    4958.000000        1987.000000  
75%              -122.288045    5502.000000        1992.000000  
max              -121.874119    6645.000000        2001.000000

In [5]:
# chek for duplicates
df.duplicated().sum()

# check for the percentage of missing values
missing_val_summary = (df.isnull().sum()/df.isnull().count())*100

display(missing_val_summary)

duration_sec               0.000000
start_time                 0.000000
end_time                   0.000000
start_station_id           0.107408
start_station_name         0.107408
start_station_latitude     0.000000
start_station_longitude    0.000000
end_station_id             0.107408
end_station_name           0.107408
end_station_latitude       0.000000
end_station_longitude      0.000000
bike_id                    0.000000
user_type                  0.000000
member_birth_year          4.506248
member_gender              4.506248
bike_share_for_all_trip    0.000000
dtype: float64

### What is the structure of your dataset?

> There are 183,412 rides made in the dataset with 16 features  (`'duration_sec', 'start_time', 'end_time', 'start_station_id', 'start_station_name', 'start_station_latitude','start_station_longitude', 'end_station_id', 'end_station_name', 'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type','member_birth_year', 'member_gender', 'bike_share_for_all_trip'`). Most variables are numeric in nature while some variables like the `start_time` and `end_time` have the wrong data types instead of the datetime.

> Also, the variables `start_station_id`, `start_station_name`, `end_station_id`, `end_station_name`, `member_birth_year` and  `member_gender` contained missing values with the following percentages:

>`start_station_id`:         0.107 <br>
`start_station_name`:        0.107 <br>
`end_station_id`:            0.107 <br>
`end_station_name`:          0.107 <br>
`member_birth_year`:         4.506 <br>
`member_gender`:             4.506 <br>

### What is/are the main feature(s) of interest in your dataset?

> Among several other informations that would warrant investigation along the data exploration, I am most interested in figuring out:
1. The time of day, day of the week that most trips are taken
2. How long does the average trip take?
3. If the answer in 1 & 2 above depend on if a user is a subscriber or a customer? 

> The main features of interest would be the `duration_sec`, `start_time`, and `user_type`. 

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> With Additional feature engineering from the main features of interest, I believe extracting the time of the day and day of the week from the start_time will help answer the number one objective listed above. It could also be expected that  the duration_sec would be able to help in answering the number two objective and then the user_type with the number three objective. Also features like `member_birth_year` and `member_gender` would greatly help support my investigations.

In [6]:
# make a copy of the original dataset to clean
df_new = df.copy()

In [7]:
# drop variables that are irrelevant to my analysis
df_new.drop(['end_time', 'start_station_id', 'start_station_latitude', 'start_station_longitude', 'end_station_id','end_station_latitude', 'end_station_longitude', 'bike_id', 
        'bike_share_for_all_trip'], axis=1, inplace=True)
df_new.head()

duration_sec                start_time   
0         52185  2019-02-28 17:32:10.1450  \
1         42521  2019-02-28 18:53:21.7890   
2         61854  2019-02-28 12:13:13.2180   
3         36490  2019-02-28 17:54:26.0100   
4          1585  2019-02-28 23:54:18.5490   

                                 start_station_name   
0  Montgomery St BART Station (Market St at 2nd St)  \
1                     The Embarcadero at Steuart St   
2                           Market St at Dolores St   
3                           Grove St at Masonic Ave   
4                               Frank H Ogawa Plaza   

                               end_station_name   user_type   
0                Commercial St at Montgomery St    Customer  \
1                            Berry St at 4th St    Customer   
2  Powell St BART Station (Market St at 4th St)    Customer   
3                        Central Ave at Fell St  Subscriber   
4                         10th Ave at E 15th St  Subscriber   

   member_birth_year member_gender  
0             1984.0          Male  
1                NaN           NaN  
2             1972.0          Male  
3             1989.0         Other  
4             1974.0          Male

In [ ]:
# assign proper data types to the start_time
df_new['start_time'] = pd.to_datetime(df_new['start_time'])

In [ ]:
# extract day of the week
df_new['day_of_week'] = pd.DatetimeIndex(df_new['start_time']).strftime('%A')
# extract the time of the day
df_new['start_hour'] = df_new['start_time'].apply(lambda time: time.hour)
df_new['time_of_day'] = 'Morning'
df_new['time_of_day'][(df_new['start_hour'] >= 12) & (df_new['start_hour'] <= 17)] = 'Afternoon'
df_new['time_of_day'][(df_new['start_hour'] >= 18) & (df_new['start_hour'] <= 23)] = 'Night'

# display the added columns
display(df_new[['day_of_week','start_hour','time_of_day']])

In [ ]:
# display the resulting dataframe
display(df_new.head())

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

I'll start by looking at the distribution of the main variables of interest: `day_of_week`, `time_of_day`, `user_type` and `duration_sec`

In [ ]:
# Get the count of trips made in each day of the week
display(df_new['day_of_week'].value_counts().sort_values(ascending=False))
weekday = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
# plot distribution of day of week
plt.figure(figsize=[8, 5])

# base_color = sns.color_palette()[0]
sns.countplot(data = df_new, x = 'day_of_week', order=weekday)
plt.title('Trips by Days of the Week')
plt.xlabel('Days')
plt.show()

It happened that the day of the week with the most trips is **Thursday** followed by Tuesday and wednesday. It could be seen that Saturday and Sunday have the least number of trips which is quite reasonable with the fact that they are not working days. Therefore, I would be investigating how the user type and gender are distributed among the days of the week. 

In [ ]:
# Get the percentage count of trips made in terms of times of the day
time_of_day_count = (df_new['time_of_day'].value_counts()/df_new['time_of_day'].count())*100
print("The percentage count of trips made in terms of times of the day(%)")
print(time_of_day_count.round(2).sort_values(ascending=False))

# plot distribution of day of week
plt.figure(figsize=[8, 5])
base_color = sns.color_palette()[0]
sns.countplot(data = df_new, x = 'time_of_day', color = base_color)
n_points = df_new.shape[0]
user_counts = df_new['time_of_day'].value_counts()
locs, labels = plt.xticks()

# print frequency on the bar chart
for loc, label in zip(locs, labels):
    count = user_counts[label.get_text()]
    pct_string = '{:0.1f}%'.format(100*count/n_points)
    plt.text(loc, count-8, pct_string, ha='center')

plt.title('Trips by Times of the Day')
plt.xlabel('Times of the Day')
plt.show()

It could be seen from the plot that the time of the day with the most trips is **Afternoon** (12:00 - 17:59 hours of the day) followed by Morning (00:00 - 11:59 hours of the day). However, the differences between the number of trips taken in the two times is very little.This is really reasonable. Noone expects so much rides in the night time. Further analysis will be done to see if this is probably influenced by user types(subscriber and customer)

In [ ]:
# Get the percentage count of user types and gender
user_type_count = (df_new['user_type'].value_counts()/df_new['user_type'].count())*100
member_gender_count = (df_new['member_gender'].value_counts()/df_new['member_gender'].count())*100

print("User Types Percentage Count (%)")
display(user_type_count.round(2).sort_values(ascending=False))
print("\nMember Gender Percentage Count (%)")
display(member_gender_count.round(2).sort_values(ascending=False))

# Attributes of the Riders: the distributions of user type and gender
fig, ax = plt.subplots(nrows=2, figsize = [8,8])
default_color = sns.color_palette()[1]
sns.countplot(data = df_new, x = 'user_type', color = default_color, ax = ax[0])
sns.countplot(data = df_new, x = 'member_gender', color = default_color, ax = ax[1])
ax[0].set_xlabel('User Type')
ax[0].set_ylabel('Count')
ax[1].set_xlabel('Member Gender')
ax[1].set_ylabel('Count')
ax[0].set_title('Distributions of User Types and Gender');

- As it would be expected in most cases, most trips (about 90%) are taken by the **subscribers** with the remaining 10% being just casual riders. In the case of larger percentage of trips by subscribers; who wouldn't want to make efficient use of their subscription?. Or in another view, the Subscribers might be using the bikes for daily commute while customers just use it for leisure activities



-  Also, almost 75% of the riders that's taken trips are **males** which is like 3 times that of the females with the remaining 2% regarded as other.

Now unto the last main variables of interest: duration_sec

In [ ]:
# start with a standard-scaled plot
binsize = 100
bins = np.arange(0, df_new['duration_sec'].max()+binsize, binsize)

plt.figure(figsize=[8, 5])
plt.hist(data = df_new, x = 'duration_sec', bins=bins)
plt.xlim(0, 3000)
plt.xlabel('Trip Duration (sec)')
plt.ylabel('Count')
plt.title('Distribution of Trip Duration')
plt.show()

The Trip duration in seconds distribution is highly right-skewed with a really long tail that it was very difficult to view the plot. I had to resolve into binning of size 100 and also set the plot's x_axis limit to 3000. I will change the trip duration variable to minute instead of second and also get to know how long the average trip take.

In [ ]:
# standardizing the trip duration variable into minutes
df_new['duration_min'] = df_new['duration_sec'] / 60
# drop the 'duration_sec" column
df_new.drop('duration_sec' , axis = 1 , inplace = True)

# show
display(df_new[['duration_min']])

In [ ]:
# print the average trip duration
average_trip_duration = np.round(df_new['duration_min'].mean(),2)
print("The time duration of average trip taken is %s minutes."%(average_trip_duration))

# trip duration distribution in log scale
log_binsize = 0.1
bins = 10 ** np.arange(0, np.log10(df_new['duration_min'].max())+log_binsize, log_binsize)
ticks =  [ 0.1 , 0.3 , 1 , 3, 10, 30, 100, 300]
labels = ['{}'.format(v) for v in ticks]

plt.figure(figsize=[8, 5])
plt.hist(data = df_new, x = 'duration_min', bins = bins)
plt.xscale('log')
plt.xticks(ticks, labels)
plt.xlabel('Trip Duration (min)')
plt.ylabel('Count')
plt.title('Log Distribution of Trip Duration')
plt.show()

When the trip duration is converted to minutes and plotted on a log-scale, the duration distribution looks highly unimodal, with the peak around 10 minutes. Interestingly, the average trip duration is around 12 minutes. 


I'll now look at the other features in the data to see if any of them hold interesting properties: `start_hour`

In [ ]:
# Get the percentage count of trips made in terms of times of the day
start_hour_count = (df_new['start_hour'].value_counts()/df_new['start_hour'].count())*100
print("The percentage count of trip starting hours (%)")
print(start_hour_count.round(2).sort_values(ascending=False))

# plot distribution of day of week
plt.figure(figsize=[8, 5])
base_color = sns.color_palette()[0]
sns.countplot(data = df_new, x = 'start_hour', color = base_color)
plt.title('Starting Hours of the trips')
plt.xlabel('Start Hour')
plt.show()

From the above barchart, it could be seen that trips are started mostly between 8-9am and 5-6pm (17-18). This is highly justifiable as these hours correspond to the hours people leave for work and home respectively.

Before closing this section, I'll take a look at the age distribution of the riders

In [ ]:
# create the age column taking the year 2019 as the current year
df_new['user_age'] = df_new['member_birth_year'].apply(lambda x: 2019 - x)
# drop the 'member_birth_year" column
df_new.drop('member_birth_year' , axis = 1 , inplace = True)
# test
display(df_new[['user_age']].head())

In [ ]:
# start with a standard-scaled plot
binsize = 1
bins = np.arange(0, df_new['user_age'].max()+binsize, binsize)

plt.figure(figsize=[8, 6])
plt.hist(data = df_new, x = 'user_age', bins=bins)
plt.xlabel('Users age (years)')
plt.ylabel('Count')
plt.title('Distribution of users age')
plt.show()

The users age distribution is highly right-skewed with as well a quite long tail. i.e most users ages ranges between 18 and 60 while there are some suprisingly old agesranging from 80 to 140. There might have been an error in data input or it could just be a case of an exception in the general expectations. 
To have a better understanding of the distibution, I would bin the ages into groups to see how age ranges are distributed.

In [ ]:
# create a new column user_age_bin from user_age group
max_age = str(int(df_new.user_age.max()))
df_new['user_age_bin'] = df_new['user_age'].apply(lambda x: '10 - 20' if 10<x<=20 else 
                                                  '21 - 30' if 20<x<=30 else
                                                  '31 - 40' if 30<x<=40 else 
                                                  '41 - 50' if 40<x<=50 else
                                                  '51 - 60' if 50<x<=60 else
                                                  '61 - 70' if 60<x<=70 else
                                                  '71 - 80' if 70<x<=80 else
                                                  '81 - '+max_age if 80<x<=float(max_age) else x)

# test
display(df_new[['user_age_bin']].head())

In [ ]:
# plot distribution of day of week
plt.figure(figsize=[8, 6])
base_color = sns.color_palette()[8]
sns.countplot(data = df_new.sort_values(['user_age_bin']), x = 'user_age_bin', color = base_color)
n_points = df_new.shape[0]
user_counts = df_new['user_age_bin'].value_counts()
locs, labels = plt.xticks()

# print frequency on the bar chart
for loc, label in zip(locs, labels):
    count = user_counts[label.get_text()]
    pct_string = '{:0.1f}%'.format(100*count/n_points)
    plt.text(loc, count-8, pct_string, ha='center')

plt.title('Distribution of age groups')
plt.xlabel('age group (years)')
plt.ylabel("count")
plt.show()

Not less than 70% of the riders ranges from age 21 to 40. i.e people of age 21 - 40 took more rides compared to all other age groups. However, I would like to investigate further on the relationship between the age groups and the duration of trips.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> The variables `duration_sec` and `user_age` (gotten from subtracting `member_birth_year` from the year the data was recorded - 2019) look very right skewed. To treat the duration_sec, I needed to convert the duration from second to minute and then used a log transform which then made the resulting distribution to be slightly unimodal with the peak around 10 minutes while the mean of the distribution was around 12 minutes. 
For the transformation of age, since the supposed outliers in the age distribution are realized to be valid years. I was able to bin/group the ages which resulted into a distribution with less skewness.

>I also checked variables like `user_type`, `member_genders` and so many other derived variables that are critical to understand the dataset and communicate the data findings. The following summarizes my findings in general:

 * The day of the week with the most trips is Thursday followed by Tuesday and wednesday. Most of the people take rides during Weekdays from Monday to Friday  compared to weekends
 * The time of the day with the most trips is Afternoon (12:00 - 17:59 hours of the day) followed by Morning (00:00 - 11:59 hours of the day)
 * Most trips (about 90%) are taken by the subscribers and would like to further investigate if this also influences the trip duration
 * Almost 75% of the riders that's taken trips are males which was like 3 times that of the females.
 * Trips are started mostly between 8-9am and 5-6pm (17-18)
 * Not less than 70% of the riders ranges from age 21 to 40, quite a youthful population, but could this relate to longer trip duration?


### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> The age distribution looked unusual, very old ages are present that seemed like outliers but I thought it was fair to let it be since they are valid. Data types of date related variables were also changed to the appropriate ones.
Also, feature engineering was performed to create features that are critical and important to my analysis. It was this feature engineering that gave birth to features like `'day_of_week', 'start_hour', 'time_of_day', 'duration_min', 'user_age'` and `'user_age_bin'`.
I as well dropped variables that aren't of interest in the dataset to tidy up the dataset for exploration.